parser/tools/generate_ast.cc

```diff
@@ -796,6 +797,69 @@ void emitNodeClassfile(ostream &out, NodeDef &node) {
+    out << '\n';
+
+    // toJSONWithLocs
+    out << "  std::string " << node.name << "::toJSONWithLocs(const core::GlobalState &gs, core::FileRef file, int tabs) {" << '\n'
+        << "    fmt::memory_buffer buf;" << '\n';
+    out << "    fmt::format_to(buf,  \"{{\\n\");\n";
+    out << "    printTabs(buf, tabs + 1);" << '\n';
+    maybeComma = "";
+    if (!node.fields.empty()) {
+        maybeComma = ",";
+    }
+    out << R"(    fmt::format_to(buf,  "\"type\" : \")" << node.name << "\\\"" << maybeComma << "\\n\");\n";
+    i = -1;
+    // Generate fields
+    for (auto &arg : node.fields) {
+        i++;
+        maybeComma = "";
+        if (i < node.fields.size() - 1) {
+            maybeComma = ",";
+        }
+        out << "    printTabs(buf, tabs + 1);" << '\n';
+        switch (arg.type) {
+            case FieldType::Name:
+                out << "    fmt::format_to(buf,  \"\\\"" << arg.name << "\\\" : \\\"{}\\\"" << maybeComma
+                    << "\\n\", JSON::escape(" << arg.name << ".data(gs)->show(gs)));\n";
+                break;
+            case FieldType::Node:
+                out << "    fmt::format_to(buf,  \"\\\"" << arg.name << "\\\" : \");\n";
+                out << "    printNodeJSONWithLocs(buf, " << arg.name << ", gs, file, tabs + 1);\n";
+                out << "    fmt::format_to(buf,  \"" << maybeComma << "\\n\");\n";
+                break;
+            case FieldType::NodeVec:
+                out << "    fmt::format_to(buf,  \"\\\"" << arg.name << "\\\" : [\\n\");\n";
+                out << "    int i = -1;" << '\n';
+                out << "    for (auto &&a: " << arg.name << ") { \n";
+                out << "      i++;\n";
+                out << "      printTabs(buf, tabs + 2);\n";
+                out << "      printNodeJSONWithLocs(buf, a, gs, file, tabs + 2);\n";
+                out << "      if (i + 1 < " << arg.name << ".size()) {\n";
+                out << "        fmt::format_to(buf,  \",\");" << '\n';
+                out << "      }" << '\n';
+                out << "      fmt::format_to(buf,  \"\\n\");\n";
+                out << "    }" << '\n';
+                out << "    printTabs(buf, tabs + 1);\n";
+                out << "    fmt::format_to(buf,  \"]" << maybeComma << "\\n\")\n;";
+                break;
+            case FieldType::String:
+                out << "    fmt::format_to(buf,  \"\\\"" << arg.name << "\\\" : \\\"{}\\\"" << maybeComma << "\\n\", "
+                    << arg.name << ");\n";
+                break;
+            case FieldType::Uint:
+                out << R"(    fmt::format_to(buf,  "\")" << arg.name << R"(\" : \"{}\")" << maybeComma << "\\n\", "
+                    << arg.name << ");\n";
+                break;
+            case FieldType::Loc:
+                out << R"(    fmt::format_to(buf,  "\"loc\" : \"{}\")" << maybeComma << "\\n\", "
+                    << "core::Loc(file, " << arg.name << ").filePosToString(gs, true));\n";
```

Comment:
We ask for the `filePosToString` with full loc to display the `endPos` of the loc

Reply:
I hate optional booleans. Can you at least assign the `true` to a variable with a descriptive name?

Other options would be to introduce an `enum class` with two options, or to make a second method that calls the first.

Reply:
I added a local variable to hold a more descriptive name.